runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/graph/GreedilyFusedExecutableStage.java

```diff
@@ -0,0 +1,223 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.beam.runners.core.construction.graph;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import java.util.ArrayDeque;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.LinkedHashSet;
+import java.util.Optional;
+import java.util.Queue;
+import java.util.Set;
+import javax.annotation.Nullable;
+import org.apache.beam.model.pipeline.v1.RunnerApi.Environment;
+import org.apache.beam.model.pipeline.v1.RunnerApi.FunctionSpec;
+import org.apache.beam.model.pipeline.v1.RunnerApi.PTransform;
+import org.apache.beam.runners.core.construction.PTransformTranslation;
+import org.apache.beam.runners.core.construction.graph.PipelineNode.PCollectionNode;
+import org.apache.beam.runners.core.construction.graph.PipelineNode.PTransformNode;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * An {@link ExecutableStage} that will greedily fuse all available {@link PCollectionNode
+ * PCollections} when it is constructed.
+ *
+ * <p>A {@link PCollectionNode} is fused into a stage if all of its consumers can be fused into the
+ * stage. A consumer can be fused into a stage if it is executed within the environment of that
+ * {@link ExecutableStage}, and receives only per-element inputs. PTransforms which consume side
+ * inputs are always the only {@link PTransformNode} at the root of a stage.
+ *
+ * <p>A {@link PCollectionNode} with consumers that execute in an environment other than a stage is
+ * materialized, and its consumers execute in independent stages.
+ */
+public class GreedilyFusedExecutableStage implements ExecutableStage {
+  // TODO: Provide a way to merge in a compatible subgraph (e.g. one where all of the siblings
+  // consume a PCollection materialized by this subgraph and can be fused into it).
+  private static final Logger LOG = LoggerFactory.getLogger(GreedilyFusedExecutableStage.class);
+
+  /**
+   * Returns an {@link ExecutableStage} where the initial {@link PTransformNode PTransform} is a
+   * Read which executes within the environment (with no input elements or {@link PCollectionNode
+   * PCollection}).
+   *
+   * <p>Once primitive Read transforms are replaced by an Impulse -> ParDo expansion, this
+   * construction should be removed.
+   */
+  public static ExecutableStage forRootTransform(
```

Comment:
Note that Dataflow replaces these with impulse transforms if beam_fn_api is enabled, https://github.com/apache/beam/blob/0cbcf4ad1db7d820c5476d636f3a3d69062021a5/runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java#L1280

All the SDK harness code expects this to be the case so we should follow this pattern here and make sure that the portable protos use impulse.

This would also make it so that each executable stage always has input pcollection.

Reply:
I've added the Impulse and ImpulseTranslation, mainly so I don't feel bad about not having any way to root a pipeline, and the URNs, and whatnot, but you should largely be able to ignore those actual changes here; the change to the default expansion and overrides in all runners to just implement Read directly should be pretty straightforward, and I'd generally prefer that in terms of it approximates the long-term goal rather than a quick "required-portability-overrides" change, but it is a small amount more work.